design/proposal-controller-code-generation.md

```diff
@@ -0,0 +1,129 @@
+
+If the OpenAPI approach turns out to make sense, we also should run the MRS generation tool against the broad surface of Provider APIs and build out a backlog of special cases and/or open issues against the core translation tool to address general issues / patterns that weren’t obvious when analyzing a particular API. For other approaches I expect there to be more edge cases and patterns to tease out which will probably spill over into phase 3.
+
+**Phase 3**: At this point we should refactor `angryjet` into the generation tool so it can operate on the MRS structure rather than process .go file outputs downstream, also deprecating `kubebuilder` in favor of directly working with the underlying `controller-tools` library. Documentation and build pipelines should be updated to reflect the new tooling. Exit from this phase will also be gated by resolving all the P0 issues with provider->MRS->boilerplate pipeline; we should have a working provider->MRS pipeline for AWS, Azure and GCP. Engage development partners to kick the tires on generation for particular providers.
+
+**Phase 4**: Clean up phase. At this point we should have a mature tool chain and process for building controller boilerplate that we can document and demo. Guide new developers on the major cloud providers to use this tool. If we feel that it is worth attempting to generate deeper CRUD automation beyond just scaffolding and boilerplate, we should at this point generate a new proposal for what that will look like in the next major version of this tool.
+
+## Related Works
+
+### Azure AutoRest (and other OpenAPI client generator tools)
+[AutoRest] is a tool built by the Azure team to generate client bindings for an OpenAPI endpoint, including client bindings underlying the Azure SDK, which is used by the Terraform Azure provider. If the tool turns out to be sufficiently generic we could explore it for generating client bindings based on OpenAPI documents for other providers, although other OpenAPI client generation tools exist like [openapi-generator]. We’ll want to take a closer look at OpenAPI code generation tools if we pursue that direction for other providers.
+
+### AWS Service Operator (ASO)
+The [pipeline architecture][aso-v2-architecture] of the next iteration of [Amazon Service Operator][amazon-aso] appears to be very similar to what we are considering, leapfrogging kubebuilder and working directly with `controller-tools` to generate controller code. The existing ASO uses CloudFormation for Provider API metadata discovery; it’s unclear from the architecture doc if they plan to continue to use CloudFormation or another API description: “The first code generation phase consumes model information from a canonical source of truth about an AWS service”.
+
+### Magic Modules
+[Magic Modules] is a Google project, written in Ruby, that does low-level code generation, translating Google’s discovery format to bindings for Terraform and Ansible.
+
+## Kubeform
```

Comment:
I personally like Kubeform very much. Not sure can we use code gen to generate CRD and controller from Terraform?

Reply:
Alternatively, it should be possible to treat Kubeform as provider and use their CRs as part of a `Composition` as a last-resort if the requested managed resource is not available in Crossplane.

Reply:
Agreed, I said it's an approach we should investigate in the `### Provider API metadata sources`. Crossplane and terraform resources fit together awkwardly in some places, like the `forProvider`/`atProvider` segmentation. Potentially terraform-derived providers could exist apart from their hand-written or openapi-generated counterparts, in a different set of repos. I'll have a stronger opinion on this after I get a chance to do some prototyping around the different metadata sources.